README.md
# Project-XOXO
FINAL
Hi this is my final project. It is a tik-tak-toe game. It is very simple and easy to use. In the game, you have to get a row of three of your shape to win. When you click run, the game starts. It tells you to click the purple box to start the game. If you want to quit at any time during the game, hit the "q" button. This game is made for kids. It is exceptionally direct and easy to use comparred to similar programs. 


changelog.md
Changlog:
# December 2, 2021: Project Inciatated
# December 20, 2021: Started working on Menu
# December 29, 2021: Menu was finished
# January 2, 2022: Started working on the game itself
# January 8, 2022: Began alpha testing
# January 10, 2022: Finished alpha testing
# January 18, 2022: Began beta testing
# January 21, 2022: Finished beta testing
# January 23, 2022: Finished work on all recomendations.
# January 23, 2022: I have fixed the bug where the game sometimes does not start.
# January 24, 2022: Finished the game part
# January 24, 2022: Projected was completed
# January 25, 2022: Project was Presented 

menu.py
# Omar Abdullat
# Computer Science 30
# January 25, 2021
# FINAL Project (Menu File)

# Here I imported Pygame.
import pygame, sys

mainClock = pygame. time.Clock()
from pygame.locals import *
pygame.init()
pygame.display.set_caption("Omar's XO Game")
screen = pygame.display.set_mode((600, 600), 0, 100)

font = pygame. font. SysFont(None, 30)


def draw_text(text, font, color, surface, x, y):
    textobj = font.render(text, 2, color)
    textrect = textobj.get_rect()
    textrect.topleft = (x, y)
    surface.blit(textobj, textrect)

# Here I made the main menu code.
def main_menu():
    while True:

        screen.fill((0, 0, 0))
        draw_text('MAIN MENU', font, (255, 255, 255), screen, 235, 25)
        draw_text('To start playing, click the purple rectrangle below.', font, (255, 255, 255), screen, 50, 70)
        mx, my = pygame.mouse.get_pos()

        button_1 = pygame.Rect(150, 150, 300, 50)
        if button_1.collidepoint((mx, my)):
            if click:
                game()

        pygame.draw.rect(screen, (22, 49, 184), button_1)

        click = False
        for event in pygame.event.get():
            if event.type == QUIT:
                pygame.quit()
                sys.exit()
            if event.type == K_ESCAPE:
                pygame.quit()
                sys.exit()
            if event.type == MOUSEBUTTONDOWN:
                if event.button == 1:
                    click = True

        pygame.display.update()
        mainClock.tick(60)

# Here is the button function and were the game is imported.
def game():
    running = True
    while running:
        screen.fill((22, 49, 184))

        import game
        for event in pygame.event.get():
            if event.type == QUIT:
                pygame.quit()
                sys.exit()
            if event.type == KEYDOWN:
                if event.key == K_ESCAPE:
                    running = False
        pygame.display.update()
        mainClock.tick(60)

main_menu()


game.py
# Omar Abdullat
# Computer Science 30
# January 25, 2021
# FINAL Project (Game File)

import pygame, sys
import numpy as np

pygame.init()

WIDTH = 600
HEIGHT = 600
LINE_WIDTH = 15
WIN_LINE_WIDTH = 15
BOARD_ROWS = 3
BOARD_COLS = 3
SQUARE_SIZE = 200
CIRCLE_RADIUS = 60
CIRCLE_WIDTH = 15
CROSS_WIDTH = 25
SPACE = 55

PURPLE = (0, 0, 0)
BG_COLOR = (22, 49, 184)
LINE_COLOR = (13, 33, 130)
CIRCLE_COLOR = (239, 231, 200)
CROSS_COLOR = (66, 66, 66)

screen = pygame.display.set_mode((WIDTH, HEIGHT))
pygame.display.set_caption("Omar's XO Game")
screen.fill(BG_COLOR)

board = np.zeros((BOARD_ROWS, BOARD_COLS))


# The Board lines are drawn here.
def draw_lines():
    # Horizontal Lines.
    pygame.draw.line(screen, LINE_COLOR, (0, 200), (600, 200), LINE_WIDTH)
    pygame.draw.line(screen, LINE_COLOR, (0, 400), (600, 400), LINE_WIDTH)
    # Vertical Lines.
    pygame.draw.line(screen, LINE_COLOR, (200, 0), (200, 600), LINE_WIDTH)
    pygame.draw.line(screen, LINE_COLOR, (400, 0), (400, 600), LINE_WIDTH)


def draw_figures():
    for row in range(BOARD_ROWS):
        for col in range(BOARD_COLS):
            if board[row][col] == 1:
                pygame.draw.circle(screen, CIRCLE_COLOR,
                                   (int(col * SQUARE_SIZE + SQUARE_SIZE // 2),
                                    int(row * SQUARE_SIZE + SQUARE_SIZE // 2)),
                                   CIRCLE_RADIUS, CIRCLE_WIDTH)
            elif board[row][col] == 2:
                pygame.draw.line(screen, CROSS_COLOR,
                                 (col * SQUARE_SIZE + SPACE,
                                  row * SQUARE_SIZE + SQUARE_SIZE - SPACE),
                                 (col * SQUARE_SIZE + SQUARE_SIZE - SPACE,
                                  row * SQUARE_SIZE + SPACE), CROSS_WIDTH)
                pygame.draw.line(
                    screen, CROSS_COLOR,
                    (col * SQUARE_SIZE + SPACE, row * SQUARE_SIZE + SPACE),
                    (col * SQUARE_SIZE + SQUARE_SIZE - SPACE,
                     row * SQUARE_SIZE + SQUARE_SIZE - SPACE), CROSS_WIDTH)


def mark_square(row, col, player):
    board[row][col] = player


def available_square(row, col):
    return board[row][col] == 0


def is_board_full():
    for row in range(BOARD_ROWS):
        for col in range(BOARD_COLS):
            if board[row][col] == 0:
                return False

    return True


def check_win(player):
    # vertical win check
    for col in range(BOARD_COLS):
        if board[0][col] == player and board[1][col] == player and board[2][
                col] == player:
            draw_vertical_winning_line(col, player)
            return True

    # horizontal win check
    for row in range(BOARD_ROWS):
        if board[row][0] == player and board[row][1] == player and board[row][
                2] == player:
            draw_horizontal_winning_line(row, player)
            return True

    # asc diagonal win check
    if board[2][0] == player and board[1][1] == player and board[0][
            2] == player:
        draw_asc_diagonal(player)
        return True

    # desc diagonal win chek
    if board[0][0] == player and board[1][1] == player and board[2][
            2] == player:
        draw_desc_diagonal(player)
        return True

    return False


def draw_vertical_winning_line(col, player):
    posX = col * SQUARE_SIZE + SQUARE_SIZE // 2

    if player == 1:
        color = CIRCLE_COLOR
    elif player == 2:
        color = CROSS_COLOR

    pygame.draw.line(screen, color, (posX, 15), (posX, HEIGHT - 15),
                     LINE_WIDTH)


def draw_horizontal_winning_line(row, player):
    posY = row * SQUARE_SIZE + SQUARE_SIZE // 2

    if player == 1:
        color = CIRCLE_COLOR
    elif player == 2:
        color = CROSS_COLOR

    pygame.draw.line(screen, color, (15, posY), (WIDTH - 15, posY),
                     WIN_LINE_WIDTH)


def draw_asc_diagonal(player):
    if player == 1:
        color = CIRCLE_COLOR
    elif player == 2:
        color = CROSS_COLOR

    pygame.draw.line(screen, color, (15, HEIGHT - 15), (WIDTH - 15, 15),
                     WIN_LINE_WIDTH)


def draw_desc_diagonal(player):
    if player == 1:
        color = CIRCLE_COLOR
    elif player == 2:
        color = CROSS_COLOR

    pygame.draw.line(screen, color, (15, 15), (WIDTH - 15, HEIGHT - 15),
                     WIN_LINE_WIDTH)


def restart():
    screen.fill(BG_COLOR)
    draw_lines()
    for row in range(BOARD_ROWS):
        for col in range(BOARD_COLS):
            board[row][col] = 0


draw_lines()

# VARIABLES

player = 1
game_over = False

# MAINLOOP

while True:
    for event in pygame.event.get():
        if event.type == pygame.QUIT:
            sys.exit()

        if event.type == pygame.MOUSEBUTTONDOWN and not game_over:

            mouseX = event.pos[0]  # x
            mouseY = event.pos[1]  # y

            clicked_row = int(mouseY // SQUARE_SIZE)
            clicked_col = int(mouseX // SQUARE_SIZE)

            if available_square(clicked_row, clicked_col):

                mark_square(clicked_row, clicked_col, player)
                if check_win(player):
                    game_over = True
                player = player % 2 + 1

                draw_figures()

        if event.type == pygame.KEYDOWN:
            if event.key == pygame.k_q:
                sys.exit()
                


    pygame.display.update()
